File: leetcode/3219.切蛋糕的最小开销II/main.cc
class Solution {
public:
    long long minimumCost(int m, int n, vector<int>& horizontalCut, vector<int>& verticalCut) {
        vector<int> rows = horizontalCut;//故意使用浅拷贝，避免多余复制
        vector<int> cols = verticalCut;//单纯想改个名，因为题干里的名字太长了
        sort(rows.begin(),rows.end());
        sort(cols.begin(),cols.end());
        long long ans = 0;
        int index1 = rows.size()-1;
        int index2 = cols.size()-1;
        int row_count = 1;
        int col_count = 1;
        while(index1>=0||index2>=0){
            int value1 = index1>=0?rows[index1]:-1;
            int value2 = index2>=0?cols[index2]:-1;
            if(value1>value2){
                ans += row_count*value1;
                col_count++;
                index1--;
            }else{
                ans += col_count*value2;
                row_count++;
                index2--;
            }
        }
        return ans;
    }
};

File: leetcode/3219.切蛋糕的最小开销II/main.py
import os
import random
import re
from cmath import inf
from functools import cache
from queue import PriorityQueue
from typing import List
import sys
import pandas as pd

sys.set_int_max_str_digits(0)


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def minimumCost(self, m: int, n: int, horizontalCut: List[int], verticalCut: List[int]) -> int:
        horizontalCut.sort()  # 默认从小到达排序
        verticalCut.sort()
        rows = horizontalCut  # 浅拷贝,不懂的百度一下
        cols = verticalCut
        ans = 0
        index1 = len(rows) - 1
        index2 = len(cols) - 1
        row_count = 1
        col_count = 1
        while index1 >= 0 or index2 >= 0:
            value1 = -1
            value2 = -1
            if index1 >= 0:
                value1 = rows[index1]
            if index2 >= 0:
                value2 = cols[index2]
            if value1 > value2:
                ans += row_count * value1
                col_count += 1
                index1 -= 1
            else:
                ans += col_count * value2
                row_count += 1
                index2 -= 1
        return ans


if __name__ == '__main__':
    s = Solution()
    s.minimumCost(3, 2, [1, 3], [5])
